File: app/identity/transport/admin_users.py
from __future__ import annotations

from collections.abc import Callable
from typing import Any, Literal

from fastapi import APIRouter, Depends, HTTPException, Path, Query
from pydantic import BaseModel, ConfigDict, Field

from app.identity.application.admin_user_diagnostics import (
    ADMIN_USER_DIAGNOSTICS_SCHEMA_VERSION,
    AdminUserDiagnosticsService,
)
PrincipalDependency = Callable[..., Any]
AdminPredicate = Callable[[Any], bool]


class AdminUserSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_id: str
    display_name: str
    status: str
    created_at: Any | None = None
    session_count: int = 0
    run_count: int = 0
    queued_run_count: int = 0
    running_run_count: int = 0
    succeeded_run_count: int = 0
    failed_run_count: int = 0
    cancelled_run_count: int = 0
    last_activity_at: Any | None = None


class AdminUserListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["ai-platform.admin-user-diagnostics.v1"] = (
        ADMIN_USER_DIAGNOSTICS_SCHEMA_VERSION
    )
    users: list[AdminUserSummaryResponse] = Field(default_factory=list)
    total: int
    offset: int
    limit: int


class AdminUserSessionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    session_id: str
    workspace_id: str
    agent_id: str
    status: str
    purpose: str
    run_count: int = 0
    failed_run_count: int = 0
    last_run_at: Any | None = None
    created_at: Any | None = None
    updated_at: Any | None = None


class AdminUserRunResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    session_id: str
    workspace_id: str
    status: str
    agent_id: str
    execution_kind: str
    skill_id: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    created_at: Any | None = None
    queued_at: Any | None = None
    started_at: Any | None = None
    finished_at: Any | None = None


class AdminUserAuditResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    audit_id: str
    actor_user_id: str | None = None
    action: str
    target_type: str
    target_id: str
    relations: list[Literal["actor", "subject", "run", "session"]]
    run_id: str | None = None
    session_id: str | None = None
    trace_id: str | None = None
    created_at: Any | None = None


class AdminUserDiagnosticsLimitsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sessions: int
    runs: int
    audit: int


class AdminUserDiagnosticsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["ai-platform.admin-user-diagnostics.v1"] = (
        ADMIN_USER_DIAGNOSTICS_SCHEMA_VERSION
    )
    user: AdminUserSummaryResponse
    sessions: list[AdminUserSessionResponse] = Field(default_factory=list)
    runs: list[AdminUserRunResponse] = Field(default_factory=list)
    audit: list[AdminUserAuditResponse] = Field(default_factory=list)
    limits: AdminUserDiagnosticsLimitsResponse


def build_admin_users_router(
    *,
    service: AdminUserDiagnosticsService,
    principal_dependency: PrincipalDependency,
    is_admin: AdminPredicate,
) -> APIRouter:
    router = APIRouter()

    @router.get("/admin/users", response_model=AdminUserListResponse)
    async def list_admin_users(
        search: str | None = Query(default=None, max_length=128),
        offset: int = Query(default=0, ge=0, le=10_000),
        limit: int = Query(default=50, ge=1, le=100),
        principal: Any = Depends(principal_dependency),
    ) -> dict[str, Any]:
        if not is_admin(principal):
            raise HTTPException(status_code=403, detail="not_ai_admin")
        normalized_search = search.strip() if search and search.strip() else None
        return await service.list_users(
            tenant_id=principal.tenant_id,
            search=normalized_search,
            offset=offset,
            limit=limit,
        )

    @router.get(
        "/admin/users/{user_id}/diagnostics",
        response_model=AdminUserDiagnosticsResponse,
    )
    async def admin_user_diagnostics(
        user_id: str = Path(
            min_length=1,
            max_length=128,
            pattern=r"^[A-Za-z0-9][A-Za-z0-9_.:@+\-]*$",
        ),
        session_limit: int = Query(default=20, ge=1, le=50),
        run_limit: int = Query(default=50, ge=1, le=100),
        audit_limit: int = Query(default=50, ge=1, le=100),
        principal: Any = Depends(principal_dependency),
    ) -> dict[str, Any]:
        if not is_admin(principal):
            raise HTTPException(status_code=403, detail="not_ai_admin")
        if ".." in user_id:
            raise HTTPException(
                status_code=400,
                detail="user_id contains unsupported characters",
            )
        result = await service.get_diagnostics(
            tenant_id=principal.tenant_id,
            user_id=user_id,
            session_limit=session_limit,
            run_limit=run_limit,
            audit_limit=audit_limit,
        )
        if result is None:
            raise HTTPException(status_code=404, detail="user_not_found")
        return result

    return router


__all__ = ["build_admin_users_router"]

File: frontend/web/src/services/api/adminUsers.ts
import { authFetch } from "./fetch";

export interface AdminUsersApiClient {
  request<T>(url: string, init?: RequestInit): Promise<T>;
}

const defaultClient: AdminUsersApiClient = {
  request: authFetch,
};

export interface AdminUserSummary {
  user_id: string;
  display_name: string;
  status: string;
  created_at?: string | null;
  session_count: number;
  run_count: number;
  queued_run_count: number;
  running_run_count: number;
  succeeded_run_count: number;
  failed_run_count: number;
  cancelled_run_count: number;
  last_activity_at?: string | null;
}

export interface AdminUserSession {
  session_id: string;
  workspace_id: string;
  agent_id: string;
  status: string;
  purpose: string;
  run_count: number;
  failed_run_count: number;
  last_run_at?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface AdminUserRun {
  run_id: string;
  session_id: string;
  workspace_id: string;
  status: string;
  agent_id: string;
  execution_kind: string;
  skill_id?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  created_at?: string | null;
  queued_at?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
}

export interface AdminUserAuditEntry {
  audit_id: string;
  actor_user_id?: string | null;
  action: string;
  target_type: string;
  target_id: string;
  relations: Array<"actor" | "subject" | "run" | "session">;
  run_id?: string | null;
  session_id?: string | null;
  trace_id?: string | null;
  created_at?: string | null;
}

export interface AdminUserListResponse {
  schema_version: "ai-platform.admin-user-diagnostics.v1";
  users: AdminUserSummary[];
  total: number;
  offset: number;
  limit: number;
}

export interface AdminUserDiagnosticsResponse {
  schema_version: "ai-platform.admin-user-diagnostics.v1";
  user: AdminUserSummary;
  sessions: AdminUserSession[];
  runs: AdminUserRun[];
  audit: AdminUserAuditEntry[];
  limits: {
    sessions: number;
    runs: number;
    audit: number;
  };
}

export interface AdminUserListOptions {
  search?: string;
  offset?: number;
  limit?: number;
}

export function buildAdminUsersUrl(options: AdminUserListOptions = {}): string {
  const params = new URLSearchParams({
    offset: String(options.offset ?? 0),
    limit: String(options.limit ?? 50),
  });
  const search = options.search?.trim();
  if (search) params.set("search", search);
  return `/api/ai/admin/users?${params.toString()}`;
}

export async function fetchAdminUsers(
  options: AdminUserListOptions = {},
  client: AdminUsersApiClient = defaultClient,
): Promise<AdminUserListResponse> {
  return client.request<AdminUserListResponse>(buildAdminUsersUrl(options), {
    method: "GET",
  });
}

export async function fetchAdminUserDiagnostics(
  userId: string,
  client: AdminUsersApiClient = defaultClient,
): Promise<AdminUserDiagnosticsResponse> {
  return client.request<AdminUserDiagnosticsResponse>(
    `/api/ai/admin/users/${encodeURIComponent(userId)}/diagnostics`,
    { method: "GET" },
  );
}

export const adminUsersApi = {
  list: fetchAdminUsers,
  diagnostics: fetchAdminUserDiagnostics,
};
